ublox_derive/tests/test.rs

```diff
@@ -1,138 +1,83 @@
-use std::convert::TryInto;
-use ublox_derive::{ubx_packet};
-use num_derive::{FromPrimitive, ToPrimitive};
-use num_traits::{FromPrimitive, ToPrimitive};
+use ublox_derive::{UbxDefineSubTypes, UbxPacketRecv, UbxPacketSend};
 
-#[ubx_packet]
-struct TestPacket {
-    field1: u32,
-    field2: u32,
-    field3: u8,
-    field4: u16,
-    field5: i32,
+trait UbxPacket {
+    const class: u8;
+    const id: u8;
+    const fixed_length: Option<u16>;
 }
 
-#[test]
-fn foo() {
-    assert_eq!(std::mem::size_of::<TestPacket>(), 15);
+#[derive(UbxPacketRecv, UbxDefineSubTypes)]
+#[ubx(class = 1, id = 2, fixed_len = 28)]
+#[repr(packed)]
+/// Geodetic Position Solution
+struct NavPosLLHRaw {
+    itow: u32,
+    #[ubx(map_type = f64, scale = 1e-7, alias = lon_degrees)]
+    lon: i32,
+    lat: i32,
+    #[ubx(map_type = f64, scale = 1e-3, alias = height_ellipsoid_meters)]
+    height_ellipsoid: i32,
+    height_msl: i32,
+    /// Horizontal Accuracy Estimate
+    horizontal_accuracy: u32,
+    vertical_accuracy: u32,
 }
 
-#[ubx_packet]
-struct TestPacket2 {
-    field1: u32,
-    field2: u8,
-    field3: u32,
+#[derive(UbxPacketRecv, UbxDefineSubTypes)]
+#[repr(packed)]
+#[ubx(class = 1, id = 3, fixed_len = 16)]
+struct NavStatusRaw {
+    itow: u32,
+    #[ubx(enum {
+	NoFix = 0,
+	DeadReckoningOnly = 1,
+	Fix2D = 2,
+	Fix3D = 3,
+	GpsDeadReckoningCombine = 4,
+	TimeOnlyFix = 5,
+    })]
+    gps_fix: u8,
+    #[ubx(bitflags {
+	GpsFixOk = bit0,
+	DiffSoln = bit1,
+	WknSet = bit2,
+	TowSet = bit3,
```

Comment:
I don't see an example offhand, is it possible to specify a field which spans multiple bits? (for example, `mapMatching` in NavStatus). In my prototyping I had used `hi:lo` syntax, which I stole from Verilog, which uses that for the same thing.

Related question, is it possible to have a bitfield map to an enum type? In the above example, I might want the enum values None, Valid, Used, and DeadReckoning.

Reply:
> is it possible to specify a field which spans multiple bits? 

Yes, in example:
> 	HotStart = mask0,

the idea for bitflags is mimic code generated by https://docs.rs/bitflags/1.2.1/bitflags/ , the most popular crate as I know for this kind of problem. In other words create strong typing for work with bits to prevent errors.

> Related question, is it possible to have a bitfield map to an enum type? In the above example, I might want the enum values None, Valid, Used, and DeadReckoning.

Not sure what you mean, enum and bitflags is divided because of it is possible to have exactly one enum value and several bitflags values. 
For bitflags you can give alias to any subset of bits via `mask` syntax.

Reply:
Ah, I see, `HotStart` isn't a field that spans multiple bits per se, it's a specific combination of setting all the bits. But each field is still a one-bit field.

How would you represent the `mapMatching` field in the `fixStat` bitfield in NavStatus? Would that be:
```
None = mask0,
Valid = mask64,
Used = mask128,
DR = mask192,
```
?

Reply:
> How would you represent the mapMatching field in the fixStat bitfield in NavStatus? Would that be:

Yep, general idea is something like this:
> Valid = mask64,
> Used = mask128,
> DR = mask192,

Though we need some proper syntax to represent 0 bits,
`None` is not `mask0`, it actually 0 in 7 and 6 bits.
May be something like this:
`None = 0_bit7 | 0_bit6`

Reply:
Having to specify the decimal numbers `192` for a bitfield type seems error-prone to me (what if we typo and type 191?)

Having the ability to directly retrieve those two bits and treat them as an enum is still type-safe.

Reply:
> Having to specify the decimal numbers 192 for a bitfield type seems error-prone to me (what > if we typo and type 191?)
> Having the ability to directly retrieve those two bits and treat them as an enum is still type-safe.

When I design this syntax I have in mind specification like:
```
0x0000 Hotstart
0x0001 Warmstart
0xFFFF Coldstart
```
so user can just copy/paste:
```rust
 Hotstart = Mask0x0000 
 Warmstart = Mask0x0001 
 Coldstart = Mask0xFFFF 
```
and that's all,
Obviously we can also support syntax like:
```
X = bit1 | bit0
```
where in specification there is notion of bits, not exact value.

To summarize:
```
X = mask IntegerConstant
X = 0_bitN | ...
X = bitN | ..
```

Reply:
Hm, yes, there are a few cases where we have a known "preset" for a bitfield. How does this interact with the rest of the flags? Is `Warmstart | Clkd` acceptable?

I think Hotstart/Warmstart/Coldstart might belong in a higher layer than raw protocol parsing. We could provide some convenience methods to make some different types of packet.

Reply:
> Is Warmstart | Clkd acceptable?

If we follow bitflags interface then yes.

> I think Hotstart/Warmstart/Coldstart might belong in a higher layer than raw protocol parsing.

What use-case have in mind? For me this is just convenient pre-defined bitset nothing more.
So I can pass:
```
nav_bbr_mask: NavBbrMask::Warmstart
```
if I want use pre-defined bitset with one more bit why not?

```
nav_bbr_mask: NavBbrMask::Warmstart | NavBbrMask::Clkd
```

Reply:
Doing this is clearly user error: `nav_bbr_mask: NavBbrMask::ColdStart | NavBbrMask::Clkd` (or at least, sloppy code on the user's part), if we allow `NavBbrMask::Warmstart | NavBbrMask::Clkd` then we have to allow both.